doc/guides/configure-a-plugin.md
# Walktrough: Configure a Plugin

*[Home](../../README.md) > [Documentation](../INDEX.md) > [Walkthroughs](./INDEX.md) > Configure a Plugin*

## Creating a configuration

The server-wide and plugin configurations are located in the root-level `config/` directory (not to be confused with `web-server/config/` for the Symfony's configuration). Each configuration file must be a YAML file.

Plugin configuration files are, by default, located in the subdirectory `config/plugins/` and named like their plugin id or in another subdirectory `config/plugins/<plugin id>/`.

## Accessing a configuration

Currently, two classes can directly access a configuration. The Settings plugin can access all configurations inside the config folder and is available everywhere, while the `BasePlugin` provides a shorthand method that makes access to the configuration of a plugin easier, but is only available in a Plugin that inherits from this base.

### Using the Settings plugin

The Settings plugin provides three public method to access a configuration:

- `get` - to read the value of one configuration from any file, or groups of configuration that can be all configurations in a file, all configurations in a YAML mapping, or all configurations of all files in a configuration sub-folder.
- `set` - to write the value of one configuration.
- `get_plugin_config` - is a shorthand to read all configurations listed in a plugin file, also available in a Twig template as `plugin_config("<plugin id>")`.

#### Syntax

```
path: [... <folders>.]<file>[:[... <mappings>.]<key>]
      [... <folders>.]<subfolder>
```

#### Values

* `<folders>` - *(Optional)* list of folders separated by periods;
* `<file>` - YAML configuration file to read;
* `<subfolder>` - folder containing the configuration files to be read;
* `<mappings>` - *(Optional)* list of period-separated keys corresponding to mappings (*a.k.a.* YAML objects);
* `<key>` - *(Optional)* key of the configuration to read.

#### Examples

```
xyz.abc:param
// Access property "param" in the file "config/xyz/abc.yaml"

xyz
// Get all configurations in "config/xyz.yaml" and in "config/xyz/"

xyz:alpha
// Access property "alpha" in the file "config/xyz.yaml"

xyz:omega.sigma

// Access child property "sigma" of the object "omega" in the file "config/xyz.yaml"
```

#### Usage

Because plugins are Symfony services, it is recommended to initialize the Settings plugin through [autowiring](https://symfony.com/doc/current/service_container/autowiring.html).

```php
use App\Plugin\Settings;

class ControllerExample
{
    // ...

                                        // Triggers autowiring
    public function my_function(int $x, Settings $settings): int
    {
        return $x + $settings->get('xyz.abc:param');
    }
}

// ...

$ctr = new ControllerExample();
$y = $ctr->my_function(10); // Notice how no second parameter is specified, whence the "auto"-wiring
// assuming 'param' is set to 3 in config/xyz/abc.yaml,
var_dump($y);   // outputs int(13)
```

### Using the `BasePlugin` shorthand

The `BasePlugin` class only provides one unique method to read and write a configuration at the same time: `BasePlugin::config`.

#### Syntax

The syntax of this function is the same as for the [previous methods](#using-the-settings-plugin), except that the function automatically precises the config path corresponding to the calling plugin. All you need to give to the function is the `key` (name) of the configuration as it is written in the YAML file and optionally its parent mappings' key or parent folders:

```php
$this->config("abc:param");
// Will read the property "param" in "config/plugins/<plugin id>/abc.yaml".

$this->config("alpha", "foo");
// Will set the property "alpha" in "config/plugins/<plugin id>.yaml" to "foo".

$this->config("omega.sigma", "bar");
// Will set the property "sigma" of object "omega" in "config/plugins/<plugin id>.yaml" to "bar".
```

Optionally, you can provide a value to assign to this configuration.

> Note: This shorthand read configuration files in the subdirectory `config/plugins/` and can only access the files inside whose name match the id of the calling plugin, or files in a subdirectory whose name match this id. To access other configurations, use the Settings plugin.

#### Usage

All plugins that inherit from `BasePlugin` can access the `BasePlugin::config` method when they have the **`Plugin` attribute** with their plugin id as parameter:

```php
<?php
// web-server/src/Plugin/PluginTest.php

namespace App\Plugin;

use App\Attribute\Plugin;
use App\Plugin\BasePlugin;

// Define this class as a plugin and precise its id
#[Plugin('test')]
class PluginTest extends BasePlugin
{
    public function calculate(float $arg): string
    {
        // Fetch the value of a setting called 'factor'
        //   in config/plugins/test.yaml
        $value = $this->config('factor');
        return $arg . ' x ' . $value . ' = ' . ($arg * $value);
    }

    public function increment(): void
    {
        // Increment the value of 'factor' by 1
        $this->config('factor', $this->config('factor') + 1);
    }
}
```

**Important:** the parent constructor uses **autowiring** to access the Settings service. This means that child classes should not override the `__construct` method. Instead, they can override the `construct` method.

If you need to override the constructor (*e.g.* to use autowiring with other services), you will need to call the parent constructor with the Settings service as parameter:

```php
use App\Attribute\Plugin;
use App\Plugin\BasePlugin;
// Import the Settings plugin
use App\Plugin\Settings;
// Import an other service
use MyService;

#[Plugin('my-plugin')]
class MyPlugin extends BasePlugin
{
    public function __construct(
        // Use autowiring in your constructor
        MyService $service,
        Settings $settings
    )
    {
        // Call the parent constructor with the Settings plugin
        parent::__construct($settings);
    }
}
```

doc/reference/yaml-syntax-for-plugin-configuration.md
# YAML syntax for plugin configuration

*[Home](../../README.md) > [Documentation](../INDEX.md) > [Reference](./INDEX.md) > YAML syntax for plugin configuration*

## YAML basics

The YAML parser used to read a plugin configuration walk through 3 layers of rules and syntax to know before writing your own configuration:

1. The universal YAML syntax (see below for documentation);
2. The Symfony's YAML component (because this parser doesn't implement some advanced YAML features, also check [the Symfony YAML component documentation page](https://symfony.com/doc/current/components/yaml.html) for both standard and php-specific features);
3. The YAML parser implemented by this project, detailed below.

## What are configurations

The Settings plugin read YAML configuration files to define the behaviors of a specific plugin or the whole server. To learn more about the configuration files' location, read the [walkthrough about configuring a plugin](../guides/configure-a-plugin.md).

A configuration is a key associated with a value. In the YAML context, every property whose value is not a mapping become a configuration. During runtime, plugins can submit queries to the Settings plugin to read or write the value of a configuration based on its key.

The Settings service provides an interface (@ `/settings[?p=<plugin id>]`) to simplify the access to these configurations using a browser. To further improve accessibility and understanding of what each configuration do, it is recommended to **document** your configuration files.

## Configuration documentation

All documentations for one property (*a.k.a.* configuration) is precised in an inline comment (starting with `#`) and the corresponding **documentation token** (preceded by `@`). Comments without any token or with unknown token are parsed as regular comments, and should not be exposed during runtime.

**All documentations are optional, and allowed in any order.** In the following, **Default** refers to the expected behavior if the token is not specified.

### Documentation token

`@name` - The text displayed to distinguish a configuration from another.

- **Syntax** : `@name <a less technical word>`
- **Default** : The name is set to the key separated in words where the key contains dashes (`-`). The first word is capitalized (the first letter is set uppercase).

`@type` - The data type of all allowed values. Mixed values are not supported: if necessary, consider using strings and deserialize them during runtime.

- **Syntax** : `@type <data type>`
- **Default** : The type is set to the data type of the parsed configuration value.

#### Allowed types

Most types can be determined automatically (*auto*) from the configuration value.

| Type     | Allowed values      | Comment                  |
| -------- | ------------------- | ------------------------ |
| string   | All textual values  | *auto*                   |
| integer  | Integer values      | *auto*                   |
| double   | Floating values     | *auto*                   |
| boolean  | True or False       | *auto*, case-insensitive |
| time     | Strings formatted `XX:YY` or `HH:XX:YY` where 0 <= `XX`, `YY` < 60 and 0 <= `HH` < 24 | Must be specified |
| select   | All values in the configuration **list** (*see the `@list` token*). | *auto*                   |

`@desc` - The text displayed below the name to precise the behavior of a configuration, its possible values or other usage notes.

- **Syntax** :
```yaml
# @desc <first line of description>[\
#       <an other line of description>]
```
*Indentation is optional.*
- **Default** : No description is displayed.

A backslash `\` can be added at the end of the line to include the next line too in the description. The backslash will be replaced by a line feed, meaning the two texts will be displayed on different lines. This is ignored if the next line starts with a valid documentation token.

If no backslash is found at the end of the line, the next line is parsed as a documentation if it starts with a valid documentation token, else as a regular comment.

`@docs` - A reference to a resource documenting this configuration or its possible values.

- **Syntax** : `@docs <URL of the documentation page>[ <text to display>]`
- **Default** : No documentation or help is provided.

A custom text can be specified instead of the default "*See more.*", following the URL and separated by a space.

`@list` - A list of possible values for this configuration. At any time this configuration's value must be in this list.

- **Syntax** : `@list <items, separated, by, commas>`
- **Default** : The configuration can take any value of its type, unless a **list** is created anyway using `@from`, the `select` type or an enum as the value.

If the provided dataset is an array, then the value is expected to be one of them.

> **Not yet implemented:**\
Else if the provided dataset is a mapping, the value of the property value is expected to be one of the key.

`@from` - (Similar to `@list`, but the list is retrieved dynamically by a function call) A function that returns a list of possible values for this configuration. At any time this configuration's value must be in this list.

- **Syntax** : `@from <service id>::<public method>`
- **Default** : The configuration can take any value of its type, unless a **list** is created anyway using `@list`, the `select` type or an enum as the value.

When parsing a configuration with this documentation, if a service has the tag `app.yaml_available` with the key that matches the `<service id>` parameter, then the method `<public method>` of this service is called, and the result become the list of possible values for this configuration.

It is expected from this method to return a mapping where values are more comprehensive than the keys. But if an array is returned, the keys become the indexes.

In the YAML configuration file or during runtime, the key is used to refer to the corresponding possible value, however the value is displayed on the interface.

## Parsing enums and constants

Instead of invoking a service using `@from`, a reference to an enum can be passed as the property value in plain YAML:

```yaml
<yaml-property>: !php/enum <Namespace\Separated\By\Backslashes\><EnumName>::<EnumConstant>
# will be parsed as an Enum object
```

If it is a backed enum, you may precise to use the value:

```yaml
language-full-name: !php/enum Example\Enum\FullNameOf::PHP->value
# will be parsed as the value of the enum
```

In the latter, the value is parsed as a string (or whatever type the enum is), but for configuration purposes, it is recommended to use an enum object (not its `->value`) so that the parser can make a list of possible values for the configuration by listing the constants in the same enum.

To read more about parsing PHP constants or enums from YAML, see the [Symfony YAML component documentation page](https://symfony.com/doc/current/components/yaml.html#parsing-php-enumerations).

## Examples

- **Basic YAML without documentation**

```yaml
max-client: 20
min-price: 0.3
contact: support@ourcompany.com
```

**Will output :**

### Max client ______________________ 20
### Min price _______________________ 0.3
### Contact _________________________ support\@ourcompany.com

___

- **Configuration name and description**

```yaml
# @desc The maximum amount of clients connected at the same time.
max-client: 20

# @name Minimum price
# @desc The minimum price of monthly subscriptions after all discounts have been applied.
min-price: 0.3

# @desc Also check our profile on our website:
# @docs https://ourcompany.com/about
contact: support@ourcompany.com
```

**Will output :**

### Max client ______________________ 20
The maximum amount of clients connected at the same time.

### Minimum price ___________________ 0.3
The minimum price of monthly subscriptions after all discounts have been applied.

### Contact _________________________ support\@ourcompany.com
Also check our profile on our website:\
*[See more](#https://ourcompany.com/about)*.

___

- **Configuration lists and multi-lines descriptions**

```yaml
# @name Color theme
# @desc The theme of the interface.
# @list blue, dark, light, orange, lime
theme: dark

# @name App Mode
# @desc The starting mode of the application.
mode: !php/enum App\Enum\AppMode::LAZY

# @name Language
# @desc The interface's language.\
#       Used for translations
# @from translate::available_languages
lang: nl
```

Assuming these PHP scripts exist:

```php
namespace App\Enum;

enum AppMode: string
{
    case LAZY = 'Lazy loading';
    case FAST = 'Fast loading';
    case VERBOSE = 'Verbose mode';
    case DEBUG = 'Debugging mode';
}
```

```php
use Symfony\Component\DependencyInjection\Attribute\AutoconfigureTag;

#[AutoconfigureTag('app.yaml_available', [ 'key' => 'translate' ])]
class MyTranslationService
{
    public function available_languages(): array
    {
        return array(
            'de' => 'Deutsch',
            'en' => 'English',
            'es' => 'Español',
            'fr' => 'Français',
            'nl' => 'Nederlands'
        );
    }
}
```

**Will output :**

### Color theme _________________ dark
The theme of the interface.

| Possible values |
| :-------------: |
| blue            |
| **dark**        |
| light           |
| orange          |
| lime            |

### App Mode ____________________ Lazy loading
The starting mode of the application.

| Possible values  |
| :--------------: |
| **Lazy loading** |
| Fast loading     |
| Verbose mode     |
| Debugging mode   |

### Language _________________________ Nederlands
The interface's language.\
Used for translations

| Possible values  |
| :--------------: |
| Deutsch          |
| English          |
| Español          |
| Français         |
| **Nederlands**   |

web-server/src/Data/Configuration.php
<?php

namespace App\Data;

use App\Exception\BaseException as Exception;
use ErrorException;
use JsonSerializable;
use Psr\Container\ContainerInterface;
use UnitEnum;

class Configuration implements JsonSerializable
{
    const string YAML_COMMENT = '#';
    const string ESCAPE_NEW_LINE = '\\';
    const string LIST_SEPARATOR = ',';
    const string METHOD_SEPARATOR = '::';
    const string PARAM_TOKEN = '@';

    const string PARAM_NAME             = 'name';
    const string PARAM_TYPE             = 'type';
    const string PARAM_DESCRIPTION      = 'desc';
    const string PARAM_DOCUMENTATION    = 'docs';
    const string PARAM_LIST             = 'list';
    const string PARAM_GET_LIST_FROM    = 'from';

    protected string $type = '';
    protected string $name = '';
    protected string $description = '';
    protected string $documentation = '';
    protected string $doc_text = '';
    // If not empty, $value MUST match one of its value (not a key!)
    protected array $possible_values = array();
    protected array $comments = array();

    public function __construct(
        protected mixed $key,
        protected mixed $value,
        array $meta,
        ContainerInterface $service_provider
    )
    {
        $waiting_new_line = false;
        foreach ($meta as $line)
        {
            $line = trim($line);
            if (!str_starts_with($line, self::YAML_COMMENT)) continue;

            $line = trim(substr($line, 1));
            if (!str_starts_with($line, self::PARAM_TOKEN))
            {
                if ($waiting_new_line)
                {
                    $waiting_new_line = str_ends_with($line, self::ESCAPE_NEW_LINE);
                    $this->description = substr($this->description, 0, -1)."\n".$line;
                }
                else
                    $this->comments[] = $line;

                continue;
            }

            $token = strtolower(substr($line, 1, strpos($line, ' ', 1) - 1));
            $content = trim(substr($line, strlen($token) + 1));
            switch ($token)
            {
                case self::PARAM_NAME:
                    $this->name = $content;
                    $waiting_new_line = false;
                    break;
                case self::PARAM_TYPE:
                    if (!$this->type && $this->type != $content) $this->type = $content;
                    else throw new Exception(
                        'Can not reassign data type of configuration "'.$this->key.'" of type "'.$this->type.'" to "'.$content.'".'
                    );
                    $waiting_new_line = false;
                    break;
                case self::PARAM_DESCRIPTION:
                    $this->description = $content;
                    if (str_ends_with($line, self::ESCAPE_NEW_LINE)) $waiting_new_line = true;
                    break;
                case self::PARAM_DOCUMENTATION:
                    [ $url, $text ] = explode(' ', $content, 2) + array('', '');

                    $this->documentation = $url;
                    $this->doc_text = $text;

                    $waiting_new_line = false;
                    break;
                case self::PARAM_LIST:
                    $list = array_map(function($item)
                    {
                        return trim($item);
                    }, explode(self::LIST_SEPARATOR, $content));
                    
                    $this->type = 'select';
                    $this->possible_values = $list;

                    $waiting_new_line = false;
                    break;
                case self::PARAM_GET_LIST_FROM:
                    try
                    {
                        [ $service_id, $method ] = explode(self::METHOD_SEPARATOR, $content, 2);
                    }
                    catch (ErrorException $th)
                    {
                        // If you are trying to refer to an enum, please use the !php/enum operator.
                        //     A whole walkthrough is available in the documentation.
                        throw new Exception(
                            'No method specified or wrong separator in the documentation token @from for configuration "'.$this->key.'".'
                        );
                    }

                    // TODO: check if $s exists
                    $s = $service_provider->get($service_id);
                    $list = $s->{$method}();
                    
                    $this->type = 'select';
                    $this->possible_values = $list;
                    $this->value = $list[$this->value];

                    $waiting_new_line = false;
                    break;
                default:
                    if ($waiting_new_line)
                    {
                        $waiting_new_line = str_ends_with($line, self::ESCAPE_NEW_LINE);
                        $this->description = substr($this->description, 0, -1)."\n".$line;
                    }
                    else
                        $this->comments[] = $line;

                    break;
            }
        }

        if (!$this->name) $this->name = ucfirst(str_replace('-', ' ', $key));

        if ($this->value instanceof UnitEnum)
        {
            // TODO: what if not a backed enum? (should be not an instance of BackedEnum)
            $this->type = 'select';
            $this->value = $value->value;
            $this->possible_values = array_combine(
                array_column($value::cases(), 'name'),
                array_column($value::cases(), 'value')
            );
        }
        else if (!$this->type) $this->type = gettype($this->value);

        if (!$this->verify($this->value))
        {
            throw new Exception(
                'Configuration "'.$this->key.'" of type "'.$this->type.'" can not be parsed or contains syntax errors.'
            );
        }
    }

    protected function verify(mixed &$value): bool
    {
        switch ($this->type)
        {
            case 'time':
                if (gettype($value) != 'string') return false;
                $div = explode(':', $value);
                
                // Check that each division of time is between 0 and 59 and contains at most two digits
                foreach ($div as $frac)
                {
                    $n = (int) $frac;
                    if (strlen($frac) > 2 || $n < 0 || $n >= 60) return false;
                }
                // Check that hour is no more than 23
                if (((int) $div[0]) >= 24)
                {
                    if (count($div) >= 3) return false;
                    
                    // Or add a new division for hours if not present
                    // Because HTML "time" inputs will parse 00:30:00 but not 30:00
                    $value = '00:'.$value;
                    $div = explode(':', $value);
                }
                
                // Check that there are at least two divisions, where each contains at least two digits
                if (count($div) < 2) $value = '00:'.str_pad($value, 2, '0', STR_PAD_LEFT);
                else if (count($div) < 4) $value = implode(':', array_map(function($frac)
                {
                    return str_pad($frac, 2, '0', STR_PAD_LEFT);
                }, $div));
                else return false;

                return true;
            case 'string':
            case 'integer':
            case 'double':
            case 'boolean':
            case 'array':
                return gettype($value) == $this->type;
            case 'select':
                return in_array($value, $this->possible_values);
            default:
                return false;
        }
    }

    public function getkey(): string
    {
        return $this->key;
    }

    public function getvalue(): mixed
    {
        return $this->value;
    }

    public function gettype(): string
    {
        return $this->type;
    }

    public function getname(): string
    {
        return $this->name;
    }

    public function getdescription(): string
    {
        return $this->description;
    }
    public function hasdescription(): bool
    {
        return $this->description != '';
    }

    public function getdocumentation(): string
    {
        return $this->documentation;
    }
    public function hasdocumentation(): bool
    {
        return $this->documentation != '';
    }

    public function getdoc_text(): string
    {
        return $this->doc_text;
    }
    public function hasdoc_text(): bool
    {
        return $this->doc_text != '';
    }

    public function getpossible_values(): array
    {
        return $this->possible_values;
    }

    public function jsonSerialize(): mixed
    {
        return $this->value;
    }
}